packages/fast-components-react-base/src/viewport-positioner/viewport-positioner.tsx

```diff
@@ -0,0 +1,1011 @@
+            xTransformOrigin,
+            right,
+            left,
+            currentHorizontalPosition: desiredHorizontalPosition,
+        };
+    };
+
+    /**
+     * Get base vertical positioning state based on desired position
+     */
+    private getBaseVerticalPositioningState = (
+        desiredVerticalPosition: ViewportPositionerVerticalPositionLabel
+    ): Partial<ViewportPositionerState> => {
+        let top: number = null;
+        let bottom: number = null;
+        let yTransformOrigin: string = "top";
+
+        switch (desiredVerticalPosition) {
+            case ViewportPositionerVerticalPositionLabel.top:
+                yTransformOrigin = "bottom";
+                bottom = this.positionerRect.height + this.anchorHeight;
+                break;
+
+            case ViewportPositionerVerticalPositionLabel.insetTop:
+                yTransformOrigin = "bottom";
+                bottom = this.positionerRect.height;
+                break;
+
+            case ViewportPositionerVerticalPositionLabel.insetBottom:
+                yTransformOrigin = "top";
+                top = -this.anchorHeight;
+                break;
+
+            case ViewportPositionerVerticalPositionLabel.bottom:
+                yTransformOrigin = "top";
+                top = 0;
+                break;
+        }
+        return {
+            yTransformOrigin,
+            top,
+            bottom,
+            currentVerticalPosition: desiredVerticalPosition,
+        };
+    };
+
+    /**
+     *  Calculate horizontal tranlation to keep positioner in view
+     */
+    private calculateHorizontalTranslate = (
+        horizontalPosition: ViewportPositionerHorizontalPositionLabel
+    ): number => {
+        if (!this.props.horizontalAlwaysInView) {
+            return 0;
+        }
+
+        let translate: number = 0;
+
+        switch (horizontalPosition) {
+            case ViewportPositionerHorizontalPositionLabel.left:
+                translate = this.viewportRect.right - this.anchorLeft;
+                translate = translate < 0 ? translate - 1 : 0;
+                break;
+
+            case ViewportPositionerHorizontalPositionLabel.insetLeft:
+                translate = this.viewportRect.right - this.anchorRight;
+                translate = translate < 0 ? translate - 1 : 0;
+                break;
+
+            case ViewportPositionerHorizontalPositionLabel.insetRight:
+                translate = this.viewportRect.left - this.anchorLeft;
+                translate = translate > 0 ? translate + 1 : 0;
+                break;
+
+            case ViewportPositionerHorizontalPositionLabel.right:
+                translate = this.viewportRect.left - this.anchorRight;
+                translate = translate > 0 ? translate + 1 : 0;
+                break;
+        }
+        return translate;
+    };
+
+    /**
+     *  Calculate vertical tranlation to keep positioner in view
+     */
+    private calculateVerticalTranslate = (
```

Comment:
prefer to add "get" here so we know there is a return value, eg. "getVerticalTranslation"

Reply:
yep